### PR TITLE
ML 1099

### PR DESCRIPTION
**Pull request by Satyam-Yadav-Infrrd**
### Jira
[ML-1099](https://infrrd.atlassian.net/browse/ML-1099)

### Description
Describe your changes, explain what you've done, how and why.
- [x] add a field userModelIds field to ClusterConfig DataClass to use that to fetch data on the basis of userModelIds
- [x] add support to create, update config with userModelIds with basic validation
- [x] Handle KeyError Exceptions which can occur while dataset preparation
- [x] update page collection with clusterInfo after training is completed.
### Creator checklist:
- [ ] The code follows the style guidelines of this project
- [ ] Run on local machine and checked the fix or the feature functionality
- [ ] Compatibility check: tested my code together with the other services
- [ ] Documentation is updated (if necessary)

#### How to test
Optional. Explain how to check your changes if it's complicated or not obvious.

### Reviewer checklist:
- [ ] I thoroughly reviewed the modified code
- [ ] I pulled the branch to my local and, if necessary, ran migrations and pip install
- [ ] Manual tests (Optional)

**As a reviewer**, if you see something not checked, please do not `+1` or `LGTM` the PR.

### CODE REVIEW

**Comment by Satyam-Yadav-Infrrd** [comment]
https://github.com/infer-ai/feedback-database-manager/pull/196
for database queries changes
`fetch_user_model_ids_entry`
`get_document_id_with_atleast_one_correction_for_fieldIds`
`get_extraction_for_document_ids_and_field_ids`
`update_cluster_info_in_page_collection`

File: training_services/DataPreparetion/BlockDataset.py [hunk 1]
```diff
@@ -78,9 +80,10 @@ def _prep_dataset(self) -> None:
         page_height = 0
         page_width = 0
 
+        prev_tag = self.other_tag
+        prev_processed_tag = self.other_tag
+
```
Comment (anuj-sadani-infrrd) [inline training_services/DataPreparetion/BlockDataset.py:85]: why this was pulled outside of the loop?

File: training_services/DataPreparetion/DataPreparation.py [hunk 2]
```diff
@@ -2,19 +2,20 @@
...
+from mongodb.services import gnn_cluster, threshold
 from tqdm import tqdm
 
 from logger import get_logger
 from training_services.DataPreparetion.DatasetPreparation import DatasetPrep
 from training_services.DataPreparetion.dataset_prep.utils import Page
 
 
-def extract_page_correction(entities_map, field_ids):
+def extract_page_correction(entities_map, field_ids, user_model_ids=None):
```
Comment (anuj-sadani-infrrd) [inline training_services/DataPreparetion/DataPreparation.py:13]: Do we expect to not have userModelId passed to the function?
if not, let us remove the default value of None

File: training_services/DataPreparetion/OCRParser.py [hunk 3]
```diff
@@ -66,33 +66,41 @@ def _parse_ocr_data(self, page: Page) -> None:
...
+                    block["endX"],
+                    block["endY"],
+                    self._page_height,
+                    self._page_width,
+                )
+                block_id = block["id"]
+                self._id2blocks[block_id] = {"bbox": bbox}
+            except KeyError as e:
+                get_logger().debug(f"Error parsing block: {e}")
+                continue
```
Comment (anuj-sadani-infrrd) [inline training_services/DataPreparetion/OCRParser.py:82]: the `continue` can be removed.

File: training_services/DataPreparetion/OCRParser.py [hunk 4]
```diff
@@ -66,33 +66,41 @@ def _parse_ocr_data(self, page: Page) -> None:
...
+                    self._page_width,
+                )
+                self._id2token[token_id] = {
+                    "bbox": bbox,
+                    "text": token_text,
+                    "parent_id": parent_id,
+                }
+            except KeyError as e:
+                get_logger().debug(f"Error parsing tokens: {e}")
+                continue
```
Comment (anuj-sadani-infrrd) [inline training_services/DataPreparetion/OCRParser.py:106]: the `continue` can be removed

File: training_services/DataPreparetion/TaggedValueParser.py [hunk 5]
```diff
@@ -26,27 +26,35 @@ def _parse_json(self, tagged_value: dict) -> None:
...
+                        tagged_token["startX"],
+                        tagged_token["startY"],
+                        tagged_token["endX"],
+                        tagged_token["endY"],
+                    )
+                    tagged_token_id = tagged_token["id"]
+                    self._id2TaggedToken[tagged_token_id] = {"bbox": bbox}
+                except KeyError as e:
+                    get_logger().info(f"Error while parsing correctedWordInfoList: {e}")
+                    continue
```
Comment (anuj-sadani-infrrd) [inline training_services/DataPreparetion/TaggedValueParser.py:40]: the continue can be removed

File: training_services/DataPreparetion/TaggedValueParser.py [hunk 6]
```diff
@@ -26,27 +26,35 @@ def _parse_json(self, tagged_value: dict) -> None:
...
+                        tagged_token["startX"],
+                        tagged_token["startY"],
+                        tagged_token["endX"],
+                        tagged_token["endY"],
+                    )
+                    tagged_token_id = tagged_token["id"]
+                    self._id2TaggedToken[tagged_token_id] = {"bbox": bbox}
+                except KeyError as e:
+                    get_logger().debug(f"Error while parsing wordInfoList: {e}")
+                    continue
```
Comment (anuj-sadani-infrrd) [inline training_services/DataPreparetion/TaggedValueParser.py:57]: the continue can be removed

File: training_services/DataPreparetion/BlockDataset.py (hunk 1 above)
Comment (Satyam-Yadav-Infrrd) [reply training_services/DataPreparetion/BlockDataset.py:85]: this was to address multi-token-multi-line entities.

File: training_services/DataPreparetion/DataPreparation.py (hunk 2 above)
Comment (Satyam-Yadav-Infrrd) [reply training_services/DataPreparetion/DataPreparation.py:13]: removed default value.

File: training_services/DataPreparetion/OCRParser.py (hunk 4 above)
Comment (Satyam-Yadav-Infrrd) [reply training_services/DataPreparetion/OCRParser.py:105]: removed

File: training_services/DataPreparetion/TaggedValueParser.py (hunk 6 above)
Comment (Satyam-Yadav-Infrrd) [reply training_services/DataPreparetion/TaggedValueParser.py]: removed

File: training_services/DataPreparetion/OCRParser.py (hunk 3 above)
Comment (Satyam-Yadav-Infrrd) [reply training_services/DataPreparetion/OCRParser.py]: removed continue

**Review by anuj-sadani-infrrd — APPROVED** [review]

**Pull request merged**